src/Language/Futhark/Syntax.hs

```diff
@@ -219,18 +224,56 @@ data AttrInfo vn
 
 -- | The elaborated size of a dimension.
 data Size
-  = -- | The size of the dimension is this name, which
-    -- must be in scope.  In a return type, this will
-    -- give rise to an assertion.
-    NamedSize (QualName VName)
-  | -- | The size is a constant.
-    ConstSize Int64
+  = -- | The size of the dimension is this expression
+    -- all non-trivial expression should have variable in scope.
+    -- In a return type, existential name don't appear in expression.
+    SizeExpr (ExpBase Info VName)
   | -- | No known size.  If @Nothing@, then this is a name distinct
     -- from any other.  The type checker should _never_ produce these
     -- - they are a (hopefully temporary) thing introduced by
     -- defunctorisation and monomorphisation.
     AnySize (Maybe VName)
-  deriving (Eq, Ord, Show)
+  deriving (Show)
+
+expSizeEq :: ExpBase NoInfo VName -> ExpBase NoInfo VName -> Bool
```

Comment:
Do we still need this?

Reply:
This are not only used in unification but also in normalization and unscopeType, they currently hold a `M.map Size VName`, but they probably could use bare cleaned exp